### PR TITLE
feat(batch): index scan & point get

### PR DESCRIPTION
**Pull request by xxchan**
## What's changed and what's your intention?
As title. Mainly added `ScanRange`, which is built in `LogicalScan::to_batch` and used in `RowSeqScan` for different types of scan.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
part of #2429

### CODE REVIEW

**Comment by wcy-fdu** [comment]
https://github.com/singularity-data/risingwave/pull/3008 merged.

**Comment by BugenZhao** [comment]
Should we make `PointGet` a separate executor? 🤔

**Comment by xxchan** [comment]
> Should we make PointGet a separate executor? 🤔

Why bother? 🤔

**Comment by xxchan** [comment]
😕
thread 'tokio-runtime-worker' panicked at 'not implemented: excluded begin key is not supported', /Users/xxchan/Projects/risingwave/src/storage/src/hummock/iterator/forward_user.rs:221:28

**Comment by xxchan** [comment]
And Unbounded in storage not handled correctly 😕

**Comment by xxchan** [comment]
finished except excluded start key 😇

File: src/storage/src/table/cell_based_table.rs [hunk 1]
```diff
@@ -332,6 +335,100 @@ impl<S: StateStore> CellBasedTable<S> {
         .await
     }
 
+    fn serialized_pk_bound(
+        &self,
+        pk_prefix: &Row,
+        next_col_bound: Bound<&Datum>,
+        is_start_bound: bool,
+    ) -> StorageResult<Bound<Vec<u8>>> {
```
Comment (fuyufjh) [inline src/storage/src/table/cell_based_table.rs:343]: So it seems unnecessary to return `StorageResult<...>`?

**Review by fuyufjh — APPROVED** [review]
LGTM!

**Comment by skyzh** [comment]
> 😕 thread 'tokio-runtime-worker' panicked at 'not implemented: excluded begin key is not supported', /Users/xxchan/Projects/risingwave/src/storage/src/hummock/iterator/forward_user.rs:221:28
> 
> @BugenZhao Any idea how to fix it?

Currently we only implemented some of the bound. If you need exclude bound, use `next_key` to convert it into an include bound.

File: src/storage/src/table/cell_based_table.rs (hunk 1 above)
Comment (xxchan) [reply src/storage/src/table/cell_based_table.rs:343]: serialize_pk returns Result

**Comment by xxchan** [comment]
> If you need exclude bound, use next_key to convert it into an include bound.

Ok I did it. Btw I also found I need to convert included end key -> excluded next key 😇

**Comment by xxchan** [comment]
image

No idea why it fails 😇

Oh my local machine, `./risedev slt -p 4566 'e2e_test/batch/range_scan.slt' -d dev` with `ci-3cn-1fe` also fails, but then I can see the correct result using `psql`.

File: src/compute/tests/integration_tests.rs [hunk 2]
```diff
@@ -436,7 +442,7 @@ async fn test_row_seq_scan() -> Result<()> {
 
     let executor = Box::new(RowSeqScanExecutor::new(
         table.schema().clone(),
-        table.iter_with_pk(u64::MAX, pk_descs).await.unwrap(),
+        ScanType::TableScan(table.iter_with_pk(u64::MAX, pk_descs).await.unwrap()),
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/compute/tests/integration_tests.rs:445]: e2e test is enough?

**Comment by xxchan** [comment]
I feel confused about whether and where to add `wait_epoch` 😇

File: src/frontend/src/utils/scan_range.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+use crate::expr::{Expr, Literal};
+
+/// See also [`ScanRangeProst`]
+#[derive(Debug, Clone)]
+pub struct ScanRange {
+    pub eq_conds: Vec<Literal>,
+    /// ((lb, inclusive), (ub, inclusive))
+    #[allow(clippy::type_complexity)]
+    pub range: Option<(Option<(Literal, bool)>, Option<(Literal, bool)>)>,
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/frontend/src/utils/scan_range.rs]: updated

File: src/storage/src/table/cell_based_table.rs (hunk 1 above)
Comment (fuyufjh) [reply src/storage/src/table/cell_based_table.rs:344]: I think `serialize_pk` should not return `StorageResult` as well...

Comment (xxchan) [reply src/storage/src/table/cell_based_table.rs:344]: Yes it returns Ok 😇

**Comment by hengm3467** [comment]
@xxchan  Does this PR contain user-facing changes? Thanks.

**Comment by xxchan** [comment]
> @xxchan Does this PR contain user-facing changes? Thanks.

This is a perfermance optimization. No behavior changes.

**Pull request merged**